File: code/samples/python/src/roles/shopping_agent_v2/shopping_agent/agent.py
"""Shopping Agent: Sub-agent hierarchy.

The hierarchy consists of consent_agent → monitoring_agent → purchase_agent.
Root agent is consent_agent; it transfers to monitoring_agent after mandate
approval. monitoring_agent transfers to purchase_agent when price constraint is
met.

Each sub-agent gets its own McpToolset instance (separate stdio connection)
per ADK best practice — sharing a single instance across agents causes
connection conflicts (see google/adk-python#712).
"""

import json
import logging
import os
import sys

from pathlib import Path
from typing import Any

from google.adk.agents import Agent
from google.adk.tools.base_tool import BaseTool
from google.adk.tools.mcp_tool import McpToolset
from google.adk.tools.mcp_tool.mcp_session_manager import StdioConnectionParams
from google.adk.tools.tool_context import ToolContext
from mcp import StdioServerParameters

from shopping_agent.mandate_tools import (
  assemble_and_sign_mandates_tool,
  check_constraints_against_mandate,
  create_checkout_presentation,
  create_payment_presentation,
  verify_checkout_receipt,
)


_AGENT_DIR = Path(__file__).resolve().parent
_AP2_ROOT = _AGENT_DIR.parent.parent
_LOG_DIR = Path(os.environ.get("LOGS_DIR", _AP2_ROOT / ".logs"))
_LOG_FILE = _LOG_DIR / "shopping-agent.log"
_LOG_DIR.mkdir(parents=True, exist_ok=True)

_shopping_logger = logging.getLogger("shopping_agent")
_shopping_logger.setLevel(logging.DEBUG)
if not _shopping_logger.handlers:
  _fh = logging.FileHandler(_LOG_FILE, mode="w", encoding="utf-8")
  _fh.setFormatter(
      logging.Formatter("%(asctime)s - %(levelname)s - %(message)s")
  )
  _shopping_logger.addHandler(_fh)

_FLOW = os.environ.get("FLOW", "x402")

_MERCHANT_SERVER = _AP2_ROOT / "merchant_agent_mcp" / "server.py"
if _FLOW == "x402":
  _CREDENTIAL_SERVER = _AP2_ROOT / "x402_credentials_provider_mcp" / "server.py"
  _MERCHANT_PAYMENT_PROCESSOR_SERVER = _AP2_ROOT / "x402_psp_mcp" / "server.py"
else:
  _CREDENTIAL_SERVER = _AP2_ROOT / "credentials_provider_mcp" / "server.py"
  _MERCHANT_PAYMENT_PROCESSOR_SERVER = (
      _AP2_ROOT / "merchant_payment_processor_mcp" / "server.py"
  )
_PROMPTS_DIR = _AGENT_DIR / "prompts"


def _make_mcp_toolset(server_path: Path, tool_filter=None) -> McpToolset:
  """Create an MCP toolset that runs a server via ``uv``.

  Each agent needs its own McpToolset instance — sharing one across
  agents causes stdio connection conflicts (google/adk-python#712).

  Args:
      server_path: The path to the MCP server script.
      tool_filter: Optional filter to select specific tools.

  Returns:
      An McpToolset instance configured to run the specified server.
  """
  env = os.environ.copy()
  if "LOGS_DIR" not in env:
    env["LOGS_DIR"] = str(_AP2_ROOT / ".logs")
  if "TEMP_DB_DIR" not in env:
    env["TEMP_DB_DIR"] = str(_AP2_ROOT / ".temp-db")

  return McpToolset(
      connection_params=StdioConnectionParams(
          server_params=StdioServerParameters(
              command=sys.executable,
              args=[server_path.name],
              cwd=str(server_path.parent),
              env=env,
          ),
          timeout=60.0,
      ),
      tool_filter=tool_filter,
  )


def _error_escalation_callback(
    tool: BaseTool,
    args: dict[str, Any],
    tool_context: ToolContext,
    tool_response: dict[str, Any],
) -> dict[str, str] | None:
  """Intercept MCP tool errors and reformat them so the LLM reliably emits.

  a structured error artifact instead of mangling or ignoring the error.

  Args:
      tool: The tool that was executed.
      args: The arguments passed to the tool.
      tool_context: The context of the tool execution.
      tool_response: The response from the tool.

  Returns:
      A dictionary containing error details and an action message if an error
      is present in `tool_response`, otherwise None.
  """
  if isinstance(tool_response, dict) and "error" in tool_response:
    code = tool_response["error"]
    msg = tool_response.get("message", str(tool_response))
    tool_name = getattr(tool, "name", str(tool))
    _shopping_logger.warning(
        "Tool %s returned error: %s – %s",
        tool_name,
        code,
        msg,
    )
    error_json = json.dumps(
        {"type": "error", "error": code, "message": msg},
        ensure_ascii=True,
    )
    return {
        "error": code,
        "message": msg,
        "action_required": (
            "STOP all processing. Emit EXACTLY this JSON as your"
            f" complete response, nothing else: {error_json}"
        ),
    }
  return None


def reset_temp_db() -> dict[str, str]:
  """Cleans up mandate files in the .temp-db directory.

  Use this tool when you want to start or restart the flow with a clean state.
  It preserves server keys.
  """
  try:
    temp_db_dir = Path(os.environ.get("TEMP_DB_DIR", _AP2_ROOT / ".temp-db"))
    if temp_db_dir.exists():
      prefixes = ["chk_", "open_chk_", "pay_", "open_pay_"]
      count = 0
      for item in temp_db_dir.iterdir():
        if item.is_file():
          if any(item.name.startswith(prefix) for prefix in prefixes):
            item.unlink()
            count += 1
      return {
          "status": "ok",
          "message": f"Removed {count} mandate files from {temp_db_dir}",
      }
    return {"status": "ok", "message": f"No temp DB found at {temp_db_dir}"}
  except Exception as e:
    return {"error": "reset_failed", "message": str(e)}


_CONSENT_INSTRUCTION = (_PROMPTS_DIR / "consent_agent.md").read_text()
_MONITORING_INSTRUCTION = (_PROMPTS_DIR / "monitoring_agent.md").read_text()
_PURCHASE_INSTRUCTION = (_PROMPTS_DIR / "purchase_agent.md").read_text()

_model = os.environ.get("AGENT_MODEL", "gemini-3.1-flash-lite-preview")

purchase_agent = Agent(
    name="purchase_agent",
    model=_model,
    description=(
        "Executes the autonomous purchase flow when price and availability"
        " satisfy the open mandates: assemble_cart, create_checkout, closed"
        " mandates, issue_payment_credential, complete_checkout,"
        " settle_with_psp."
    ),
    instruction=_PURCHASE_INSTRUCTION,
    output_key="purchase_result",
    tools=[
        check_constraints_against_mandate,
        create_checkout_presentation,
        create_payment_presentation,
        verify_checkout_receipt,
        _make_mcp_toolset(_MERCHANT_SERVER),
        _make_mcp_toolset(_CREDENTIAL_SERVER),
        _make_mcp_toolset(
            _MERCHANT_PAYMENT_PROCESSOR_SERVER,
            tool_filter=lambda tool, ctx=None: tool.name != "initiate_payment",
        ),
    ],
    after_tool_callback=_error_escalation_callback,
)

monitoring_agent = Agent(
    name="monitoring_agent",
    model=_model,
    description=(
        "Holds open mandates and monitors item price and availability via"
        " check_product. Transfers to purchase_agent when price is within the"
        " mandate and the merchant reports the item as available."
    ),
    instruction=_MONITORING_INSTRUCTION,
    output_key="monitoring_result",
    tools=[
        check_constraints_against_mandate,
        _make_mcp_toolset(_MERCHANT_SERVER),
    ],
    sub_agents=[purchase_agent],
    after_tool_callback=_error_escalation_callback,
)

consent_agent = Agent(
    name="consent_agent",
    model=_model,
    description=(
        "Handles drop/budget dialogue, product search, item selection, and"
        " open-mandate signing. Transfers to monitoring_agent after"
        " mandate_approved or on 'Check price now'."
    ),
    instruction=_CONSENT_INSTRUCTION,
    output_key="consent_result",
    tools=[
        reset_temp_db,
        assemble_and_sign_mandates_tool,
        _make_mcp_toolset(_MERCHANT_SERVER),
    ],
    sub_agents=[monitoring_agent],
    after_tool_callback=_error_escalation_callback,
)

root_agent = consent_agent

File: code/samples/typescript/src/roles/shopping-agent-v2/agent.ts
/**
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Shopping Agent v2 — Human-Not-Present.
 *
 * Mirrors the Python `shopping_agent_v2` sub-agent hierarchy:
 *   consent_agent (root) -> monitoring_agent -> purchase_agent (leaf).
 *
 * - consent_agent: drop/budget dialogue + open-mandate signing, then transfers
 *   to monitoring.
 * - monitoring_agent: watches price/availability, transfers to purchase when the
 *   open-mandate constraints are satisfied and the item is available.
 * - purchase_agent: runs the autonomous purchase pipeline (assemble cart ->
 *   create checkout -> closed-mandate presentations -> issue payment credential
 *   -> complete checkout -> verify receipts).
 *
 * Each agent owns its OWN set of MCPToolset instances (separate stdio
 * connections) per ADK best practice — sharing a single toolset across agents
 * causes stdio connection conflicts (see google/adk-python#712). Each MCP
 * toolset launches the corresponding `*-mcp/server.ts` as a stdio subprocess.
 */

import { LlmAgent, MCPToolset } from '@google/adk';
import type { BaseTool } from '@google/adk';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

import {
  assembleAndSignMandatesTool,
  checkConstraintsAgainstMandateTool,
  createCheckoutPresentationTool,
  createPaymentPresentationTool,
  verifyCheckoutReceiptTool,
  resetTempDbTool,
} from './mandate-tools.js';
import { CONSENT_INSTRUCTION } from './prompts/consent.js';
import { MONITORING_INSTRUCTION } from './prompts/monitoring.js';
import { PURCHASE_INSTRUCTION } from './prompts/purchase.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROLES_DIR = path.resolve(__dirname, '..');

const MERCHANT_SERVER = path.join(ROLES_DIR, 'merchant-agent-mcp/server.ts');
const CREDENTIAL_SERVER = path.join(ROLES_DIR, 'credentials-provider-mcp/server.ts');
const PSP_SERVER = path.join(ROLES_DIR, 'merchant-payment-processor-mcp/server.ts');

const MCP_ENV = {
  ...process.env,
  LOGS_DIR: process.env.LOGS_DIR ?? path.resolve(ROLES_DIR, '../../.logs'),
  TEMP_DB_DIR: process.env.TEMP_DB_DIR ?? path.resolve(ROLES_DIR, '../../.temp-db'),
} as Record<string, string>;

/**
 * Build a fresh MCPToolset that launches an `*-mcp/server.ts` over stdio.
 *
 * A new instance is created per call (and per agent) on purpose: sharing one
 * toolset across agents reuses the same stdio session and conflicts.
 *
 * @param serverEntry Absolute path to the MCP server entrypoint.
 * @param toolFilter Optional allowlist of tool names or a predicate.
 * @returns A configured MCPToolset.
 */
function makeMcpToolset(
  serverEntry: string,
  toolFilter?: string[] | ((tool: BaseTool) => boolean),
): MCPToolset {
  return new MCPToolset(
    {
      type: 'StdioConnectionParams',
      serverParams: {
        command: 'npx',
        args: ['tsx', serverEntry],
        env: MCP_ENV,
      },
      timeout: 60_000,
    },
    toolFilter,
  );
}

const MODEL = process.env.AGENT_MODEL ?? 'gemini-2.5-flash';

/**
 * Reformat any MCP/mandate tool error into a structured artifact and escalate
 * so the LLM reliably stops and emits exactly {type:'error', error, message}.
 *
 * Returning a replacement object becomes the tool result the model sees, and
 * setting `context.actions.escalate` halts further sub-agent processing.
 */
const errorEscalationCallback = ({
  tool,
  response,
  context,
}: {
  tool: BaseTool;
  args: Record<string, unknown>;
  response: Record<string, unknown>;
  context: { actions: { escalate?: boolean } };
}): Record<string, unknown> | undefined => {
  if (response && typeof response === 'object' && 'error' in response) {
    const error = (response as { error: unknown }).error;
    const message = 'message' in response ? (response as { message: unknown }).message : String(response);
    context.actions.escalate = true;
    const errorJson = JSON.stringify({ type: 'error', error, message });
    return {
      error,
      message,
      action_required:
        `STOP all processing for tool "${tool.name}". Emit EXACTLY this JSON ` +
        `as your complete response, nothing else: ${errorJson}`,
    };
  }
  return undefined;
};

export const purchaseAgent = new LlmAgent({
  name: 'purchase_agent',
  model: MODEL,
  description:
    'Executes the autonomous purchase flow once price and availability satisfy ' +
    'the open mandates: assemble_cart, create_checkout, closed mandates, ' +
    'issue_payment_credential, complete_checkout, and receipt verification.',
  instruction: PURCHASE_INSTRUCTION,
  outputKey: 'purchase_result',
  tools: [
    checkConstraintsAgainstMandateTool,
    createCheckoutPresentationTool,
    createPaymentPresentationTool,
    verifyCheckoutReceiptTool,
    makeMcpToolset(MERCHANT_SERVER),
    makeMcpToolset(CREDENTIAL_SERVER),
    // The purchase agent must NOT settle directly via the PSP; filter out
    // initiate_payment (mirrors the Python tool_filter lambda).
    makeMcpToolset(PSP_SERVER, (tool) => tool.name !== 'initiate_payment'),
  ],
  afterToolCallback: errorEscalationCallback,
});

export const monitoringAgent = new LlmAgent({
  name: 'monitoring_agent',
  model: MODEL,
  description:
    'Holds the open mandates and monitors item price and availability via ' +
    'check_product. Transfers to purchase_agent when the price is within the ' +
    'mandate and the merchant reports the item as available.',
  instruction: MONITORING_INSTRUCTION,
  outputKey: 'monitoring_result',
  tools: [checkConstraintsAgainstMandateTool, makeMcpToolset(MERCHANT_SERVER)],
  subAgents: [purchaseAgent],
  afterToolCallback: errorEscalationCallback,
});

export const consentAgent = new LlmAgent({
  name: 'consent_agent',
  model: MODEL,
  description:
    'Handles drop/budget dialogue, item selection, and open-mandate signing. ' +
    'Transfers to monitoring_agent after the mandates are approved or on ' +
    '"Check price now".',
  instruction: CONSENT_INSTRUCTION,
  outputKey: 'consent_result',
  tools: [
    resetTempDbTool,
    assembleAndSignMandatesTool,
    checkConstraintsAgainstMandateTool,
    makeMcpToolset(MERCHANT_SERVER),
  ],
  subAgents: [monitoringAgent],
  afterToolCallback: errorEscalationCallback,
});

export const shoppingAgentV2 = consentAgent;

export { shoppingAgentV2 as rootAgent };
